### ISSUE TITLE
使用js解题遇到无法读取数据的问题

### ISSUE THREAD

**Issue by liuyingbin1922** [zhblue/hustoj#678]
判题机已经通过源码安装NODEJS ， 并且 可以编译js文件，但是使用js 提交代码，出现答案错误 ， 错误界面是:
========[test.out]=========
Expected						      |	Yours
6							      <
30							      <
==============================
========[sample.out]=========
Expected						      |	Yours
6							      <
30							      <
==============================
使用debug 模式进行调试 ，发现error.out 是空， 想知道这是什么情况 ？ 一直没有解决这一问题。

**Comment by liuyingbin1922** [comment]
请大佬指点

**Comment by MX-Qulin** [comment]
你的程序没有输出

试试看提交仅有输出的代码有没有效果
```
```

**Comment by zhblue** [comment]
具体发行版给一下，这个估计是缺少.so文件和相关权限的原因，需要chroot进去手工跟踪

**Comment by liuyingbin1922** [comment]
> 具体发行版给一下，这个估计是缺少.so文件和相关权限的原因，需要chroot进去手工跟踪
是在 Ubuntu 14.04.6 LTS 系统上进行安装的 ， 单纯的输出console.log() 也是不起作用的。

**Comment by liuyingbin1922** [comment]
> > 具体发行版给一下，这个估计是缺少.so文件和相关权限的原因，需要chroot进去手工跟踪
> > 是在 Ubuntu 14.04.6 LTS 系统上进行安装的 ， 单纯的输出console.log() 也是不起作用的。

还想问一下，该怎么调试和修改这一个问题 ？

**Comment by zhblue** [comment]
用的hustoj是最新版吗？我刚好有一个14.04.的系统，测试可以跑过A+B

**Comment by liuyingbin1922** [comment]
> 用的hustoj是最新版吗？我刚好有一个14.04.的系统，测试可以跑过A+B

不是 ， 用的是18年的代码 ， 用这一版的代码在公司搞OJ比赛好久了 ， 觉得还是在这一版本上解决问题 ，不更新到最新了。大佬 ， 这一版本的该怎么解决这一问题？

**Comment by zhblue** [comment]
这样的话，建议直接更新两个东西就好
1、judge_client.cc中的[copy_js_runtime](https://github.com/zhblue/hustoj/blob/master/trunk/core/judge_client/judge_client.cc#L2143)
2、okcalls32.h 或 okcalls64.h ，取决于你系统的平台是32还是64。

**Comment by zhblue** [comment]
解决类似问题的方法是在运行debug模式的judge_client后，来到对应的run0/log下，进行sudo chroot . 
获得一个类似运行时沙箱的文件系统，然后尝试运行你的程序，如:
         ./nodejs Main.js < data.in > user.out
然后根据报错来判断所缺的文件，然后修改copy_js_runtime，进行补充

**Comment by Chenjiayuan195** [comment]
在运行debug模式后，并未在run0中找到log，需要怎么办

**Comment by zhblue** [comment]
请问你判题机的普通用户账号登录名不会取了个judge吧

**Comment by Chenjiayuan195** [comment]
你好，不是取名为judge的

**Comment by zhblue** [comment]
请描述一下你的debug模式怎么运行的，完整命令行是什么

**Comment by Chenjiayuan195** [comment]
sudo judge_client 30236 0 /home/judge/ debug 
30236是题目编号

**Comment by zhblue** [comment]
> sudo judge_client 30236 0 /home/judge/ debug
> 30236是题目编号

看起来应该是正常评测把你的log目录清掉了
试试
sudo judge_client 30236 10 /home/judge/ debug
然后去 /home/judge/run10/log
前提是你的OJ_RUNNING小于等于10

**Comment by Chenjiayuan195** [comment]
前提是你的OJ_RUNNING小于等于10
这个的意思run的个数要小于10吗？我看现在里面有到run40

**Comment by Chenjiayuan195** [comment]
我目前看到run10下的error.out中存在
/nodejs: error while loading shared libraries: libcares.so.2: cannot open shared object file: No such file or directory
不知道是不是之前的log，但我查找了
execute_cmd("/bin/cp /usr/lib/x86_64-linux-gnu/libcares.so.2  %s/lib/x86_64-linux-gnu/",work_dir)
该so文件存在，想请问一下，这里该如何修改一下

**Comment by zhblue** [comment]
这有可能是libcares.so.2 这个文件还依赖别的.so，但是没有一并复制过来的原因
可以用ldd /usr/lib/x86_64-linux-gnu/libcares.so.2 看一下依赖关系，把确实的文件用类似的方式复制到运行目录下

**Comment by Chenjiayuan195** [comment]
依赖链都是有的，我看了error.out没有打印了，但diff.out还是错的，我看run10下也同样不存在log文件夹

**Comment by zhblue** [comment]
看一下judge.conf里面OJ_RUNNING的设置是多少

**Comment by Chenjiayuan195** [comment]
OJ_RUNNING=1

**Comment by zhblue** [comment]
那就有点不对了，除非你曾经运行过judge_client xxxxx 40 /home/judge    否则不应该出现run40

**Comment by zhblue** [comment]
由于你用的老版本，可能还要把copy_shell_runtime这个函数也一并更新一下

**Comment by Chenjiayuan195** [comment]
应该是运行过judge_client xxxxx 40 /home/judge的，copy_shell_runtime这个需要更新什么呢

**Comment by Chenjiayuan195** [comment]
还想请教一下，我看每次提交会有不同的判题机来执行，这个判题机怎么定位呢

**Comment by zhblue** [comment]
用最新的copy_shell_runtime代码去覆盖你的老版本，然后重新编译judge_client，覆盖到/usr/bin去

**Comment by zhblue** [comment]
你自己部署了哪些判题机，自己不清楚么？

**Comment by Chenjiayuan195** [comment]
我现在又遇到一个问题，就是运行debug之后会error.out报错
```
Runtime Error:Segmentation fault
Runtime Error:Segmentation fault
```
想请教一下作者这个问题应该怎么定位

**Comment by zhblue** [comment]
段错误正常情况下多意味着程序指针越界了，debug的话可能是参数给的不对或者数据库有问题
